pkg/service/keystore/storage.go

```diff
@@ -48,43 +56,107 @@ const (
+}
+
+type wrappedEncryptor struct {
+	tink.AEAD
+}
+
+func (w wrappedEncryptor) Encrypt(_ context.Context, plaintext, contextData []byte) ([]byte, error) {
+	return w.AEAD.Encrypt(plaintext, contextData)
+}
+
+type wrappedDecryptor struct {
+	tink.AEAD
+}
+
+func (w wrappedDecryptor) Decrypt(_ context.Context, ciphertext, contextInfo []byte) ([]byte, error) {
+	return w.AEAD.Decrypt(ciphertext, contextInfo)
+}
+
+const (
+	gcpKMSScheme = "gcp-kms"
+	awsKMSScheme = "aws-kms"
+)
+
+func NewEncryption(db storage.ServiceStorage, cfg config.KeyStoreServiceConfig) (Encryptor, Decryptor, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if strings.HasPrefix(cfg.MasterKeyURI, gcpKMSScheme) || strings.HasPrefix(cfg.MasterKeyURI, awsKMSScheme) {
+		return NewExternalEncryptor(ctx, cfg)
+	}
+
+	// First, generate a service key
+	serviceKey, serviceKeySalt, err := GenerateServiceKey(cfg.ServiceKeyPassword)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not decode service key")
+		return nil, nil, sdkutil.LoggingErrorMsg(err, "generating service key")
 	}
-	bolt := &Storage{db: db, serviceKey: keyBytes}
 
-	// first, store the service key
-	if err = bolt.storeServiceKey(context.Background(), key); err != nil {
-		return nil, errors.Wrap(err, "could not store service key")
+	key := ServiceKey{
+		Base58Key:  serviceKey,
+		Base58Salt: serviceKeySalt,
 	}
-	return bolt, nil
+	if err := storeServiceKey(ctx, db, key); err != nil {
+		return nil, nil, err
+	}
+	return &encryptor{db}, &decryptor{db}, nil
+}
+
+func NewExternalEncryptor(ctx context.Context, cfg config.KeyStoreServiceConfig) (Encryptor, Decryptor, error) {
+	var client registry.KMSClient
+	var err error
+	if strings.HasPrefix(cfg.MasterKeyURI, gcpKMSScheme) {
+		client, err = gcpkms.NewClientWithOptions(ctx, cfg.MasterKeyURI, option.WithCredentialsFile(cfg.KMSCredentialsPath))
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "creating gcp kms client")
+		}
+	} else if strings.HasPrefix(cfg.MasterKeyURI, awsKMSScheme) {
+		client, err = awskms.NewClientWithCredentials(cfg.MasterKeyURI, cfg.KMSCredentialsPath)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "creating aws kms client")
+		}
+	}
+	registry.RegisterKMSClient(client)
+	dek := aead.XChaCha20Poly1305KeyTemplate()
+	kh, err := keyset.NewHandle(aead.KMSEnvelopeAEADKeyTemplate(cfg.MasterKeyURI, dek))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "creating keyset handle")
+	}
+	a, err := aead.New(kh)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "creating aead from key handl")
+	}
+	return wrappedEncryptor{a}, wrappedDecryptor{a}, nil
 }
 
 // TODO(gabe): support more robust service key operations, including rotation, and caching
-func (kss *Storage) storeServiceKey(ctx context.Context, key ServiceKey) error {
+func storeServiceKey(ctx context.Context, db storage.ServiceStorage, key ServiceKey) error {
 	keyBytes, err := json.Marshal(key)
```

Comment:
can check if the key is empty

Reply:
I'd rather leave as it, as this was refactoring. I don't want to change behaviors if possible.